CMakeLists.txt

```diff
@@ -24,119 +24,126 @@ enable_testing()
-			# If not, its value will be "b2-NOTFOUND"
-			b2
-
-			# Name of the file to look for depending on the host OS
-			b2.exe
-
-			PATHS "${BOOST_SOURCEDIR}"
-			
-			# Prevents the function from using CMake cache to retrieve the path and
-			# avoid errors in case b2 was deleted between cache generation events
-			NO_CMAKE_PATH
-		)
-	else()
-		find_file(
-			# Variable which holds the path to b2 if it's found
-			# If not, its value will be "b2-NOTFOUND"
-			b2
-
-			# Name of the file to look for depending on the host OS
-			b2
-
-			PATHS "${BOOST_SOURCEDIR}"
-		)
-	endif()
-
-	if (${b2} STREQUAL "b2-NOTFOUND")
-		boost_configure()
-	endif()
-
-	boost_compile()
-	boost_set_paths()
+option(VITA3K_FORCE_CUSTOM_BOOST "Force Vita3K build process to use the Boost version included with the repository" OFF)
+option(VITA3K_FORCE_SYSTEM_BOOST "Force Vita3K build process to use the Boost version in the system and CMake's default paths and ignore the Boost version included with Vita3K" OFF)
+
+# If build process isn't set to forcefully use system Boost
+if(NOT VITA3K_FORCE_SYSTEM_BOOST)
+    # find_package(Boost ...) setting
+    set(Boost_USE_STATIC_LIBS ON)
+
+    # First, try to find Boost without any hints (system Boost)
+    if(NOT VITA3K_FORCE_CUSTOM_BOOST)
+        find_package(Boost COMPONENTS filesystem system program_options QUIET)
+    endif()
+
+    # If system Boost hasn't been found, then enable hints to find custom Boost
+    if (NOT Boost_FOUND)
+        message(STATUS "A Boost distribution that meets the requirements needed for the project hasn't been found in your system. Falling back to custom Boost distribution...")
+    
+        # Set path hints for Boost search inside the repository folder and assist in compilation if needed
+        set(BOOST_SOURCEDIR "${CMAKE_SOURCE_DIR}/external/boost")       # For internal use
+        set(BOOST_INSTALLDIR "${CMAKE_BINARY_DIR}/external/boost")      # For internal use
+        set(BOOST_ROOT "${BOOST_SOURCEDIR}")                            # find_package(Boost ...) hint
+        set(BOOST_INCLUDEDIR "${BOOST_SOURCEDIR}")                      # find_package(Boost ...) hint
+        set(BOOST_LIBRARYDIR "${BOOST_INSTALLDIR}/lib")                 # find_package(Boost ...) hint
+
+        # Find Boost again to check for a existing compilation of the custom distribution
+        find_package(Boost COMPONENTS filesystem system program_options QUIET)
+
+        # If no build of the custom distribution is found, compile it
+        if(NOT Boost_FOUND)
+            message(STATUS "No existing custom distribution of Boost has been found in ${BOOST_INSTALLDIR}. Proceeding to compile Boost...")
+
+            # Set compilation toolset for b2 and Boost builds
+            if (WIN32)
+                if(CMAKE_GENERATOR_TOOLSET STREQUAL "clangcl")
+                    set(BOOST_TOOLSET "clang-win")
+                else()
+                    set(BOOST_TOOLSET msvc)
+                endif()
```

Comment:
I was not good in cmake, so CMAKE_CXX_COMPILER_ID may be checked here too.

Reply:
For MSBuild (Visual Studio) is fine to check against `CMAKE_GENERATOR_TOOLSET`, but yeah it might be better to check against `CMAKE_CXX_COMPILER_ID` so that it also works for builds that don't use MSBuild.